packages/http-crawler/src/internals/file-download.ts

```diff
@@ -0,0 +1,219 @@
+ *
+ * We can use the `preNavigationHooks` to adjust `gotOptions`:
+ *
+ * ```
+ * preNavigationHooks: [
+ *     (crawlingContext, gotOptions) => {
+ *         // ...
+ *     },
+ * ]
+ * ```
+ *
+ * New requests are only dispatched when there is enough free CPU and memory available, using the functionality provided by the [AutoscaledPool](https://crawlee.dev/api/core/class/AutoscaledPool) class. All [AutoscaledPool](https://crawlee.dev/api/core/class/AutoscaledPool) configuration options can be passed to the `autoscaledPoolOptions` parameter of the `FileCrawler` constructor. For user convenience, the `minConcurrency` and `maxConcurrency` [AutoscaledPool](https://crawlee.dev/api/core/class/AutoscaledPool) options are available directly in the `FileCrawler` constructor.
+ *
+ * ## Example usage
+ *
+ * ```ts
+ * const crawler = new FileDownloader({
+ *     requestHandler({ body, request }) {
+ *         writeFileSync(request.url.replace(/[^a-z0-9\.]/gi, '_'), body);
+ *     },
+ * });
+ *
+ * await crawler.run([
+ *     'http://www.example.com/document.pdf',
+ *     'http://www.example.com/sound.mp3',
+ *     'http://www.example.com/video.mkv',
+ * ]);
+ * ```
+*/
+export class FileDownload extends HttpCrawler<FileDownloadCrawlingContext> {
+    private streamHandler?: StreamHandler;
+
+    constructor(options: FileDownloadCrawlerOptions = {}) {
+        const { streamHandler } = options;
+        delete options.streamHandler;
+        super(options);
+
+        this.streamHandler = streamHandler;
+        if (this.streamHandler) {
+            this.requestHandler = this.streamRequestHandler;
+        }
+
+        (this as any).supportedMimeTypes = new Set(['*/*']);
+    }
+
+    protected override async _runRequestHandler(context: FileDownloadCrawlingContext) {
+        if (this.streamHandler) {
+            context.request.skipNavigation = true;
+        }
+
+        await super._runRequestHandler(context);
+    }
+
+    private async streamRequestHandler(context: FileDownloadCrawlingContext) {
+        const { log, request: { url } } = context;
+
+        const { gotScraping } = await import('got-scraping');
+
+        const stream = gotScraping.stream({
+            url,
+            timeout: { request: undefined },
+            proxyUrl: context.proxyInfo?.url,
+            isStream: true,
+        });
+
+        let pollingInterval: NodeJS.Timeout | undefined;
+
+        const cleanUp = () => {
+            clearInterval(pollingInterval!);
+            stream.destroy();
+        };
+
+        const downloadPromise = new Promise<void>((resolve, reject) => {
+            pollingInterval = setInterval(() => {
+                const { total, transferred } = stream.downloadProgress;
+
+                if (transferred > 0) {
+                    log.info(
+                        `Downloaded ${transferred} bytes of ${total ?? 0} bytes from ${url}.`,
+                    );
+                }
+            }, 1000);
```

Comment:
Note for discussion: do we want some default implementation like this here, or should we give users the opportunity to register their own "download progress" hooks?

Reply:
Either way, this is currently only in the "streaming" version of the download - and moving it to the navigation inside `HttpCrawler` might be a bit of a pain. It adds valuable info, though.

Reply:
I'd stick with the default log for now. The 1s interval could swamp the logs pretty quickly though.